README.md

```diff
@@ -1 +1,571 @@
+  email: String!
+  password: String!
+}
+```
+
+But then imagine that it is decided that users should also be able to register with a phone number instead of an email address. The intuitive approach might be to make `email` optional and add an optional `phone` field:
+
+```graphql
+input RegisterUserInput {
+  email: String
+  phone: String
+  password: String!
+}
+```
+
+This is bad because it allows mutation requests with neither `email` or `phone` and with both `email` and `phone`, neither of which are allowed. What we should do is create a new mutation for this method of registration and, optionally, deprecate the old method.
+
+```graphql
+type Mutation {
+  registerUser(input: RegisterUserInput!): RegisterUserPayload! @deprecated(reason: "Use `registerUserWithEmail`.")
+  registerUserWithEmail(input: RegisterUserWithEmailInput!): RegisterUserWithEmailPayload!
+  registerUserWithPhone(input: RegisterUserWithPhoneInput!): RegisterUserWithPhonePayload!
+}
+
+input RegisterUserInput {
+  email: String! @deprecated(reason: "Use `registerUserWithEmail` mutation.")
+  password: String! @deprecated(reason: "Use `registerUserWithEmail` mutation.")
+}
+
+input RegisterUserWithEmailInput {
+  email: String!
+  password: String!
+}
+
+input RegisterUserWithPhoneInput {
+  email: String!
+  password: String!
+}
+```
+
+This makes it impossible to represent obviously incorrect inputs.
+
+This would be easier if GraphQL had input unions, which might be added in the future.
+
+#### Output
+
+All mutations should return a payload output object that is specific to that mutation and named as such.
+
+```graphql
+type Mutation {
+  registerUser(input: RegisterUserInput!): RegisterUserPayload!
+}
+```
+
+All mutation payloads should implement the following interface:
+
+```graphql
+interface MutationPayload {
+  query: Query! # the root query object
+  # mutation payloads should also return a `result` object
+}
+```
+
+Mutation payloads should always return the root query object so that clients can query from the root, regardless of what the mutation response is otherwise. This can be useful for a variety of reasons, but the primary reason is that stateful clients like Relay and Apollo need to incorporate the mutation changes into the state but do not know how to by default because the effects of the mutation are not known ahead of time. Clients can use the query root to query for changed data which should replace stale local data.
+
+Additionally, Mutation payloads should return a `result` union. Please read [this article about result unions](https://medium.com/@sachee/200-ok-error-handling-in-graphql-7ec869aec9bc). Unfortunately, GraphQL's type system cannot express that `MutationPayload` must have a field called `result` and the value must be non-scalar. If GraphQL's type system could express this interface, it might look like this:
+
+```graphql
+interface MutationPayload {
+  query: Query!
+  result: Object! # impossible to express currently
```

Comment:
Not recommending this, but wouldn't a union type of every possible result express this?

Reply:
Sort of, I think. It seems possible, assuming the schema is generated, but the actual result types would be more specific and I think GraphQL current requires interface field implementations to be an exact match, not just a subset.